sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java

```diff
@@ -1010,6 +1037,118 @@ public boolean getValidate() {
+      this.coder = coder;
+    }
+
+    @Override
+    public FileBasedSink.FileBasedWriteOperation<T> createWriteOperation(PipelineOptions options) {
+      return new BigQueryWriteOperation<>(
+          this, jobIdToken, jsonTable, jsonSchema, writeDisposition, createDisposition, coder);
+    }
+
+    private static class BigQueryWriteOperation<T> extends FileBasedWriteOperation<T> {
+      private final String jobIdToken;
+      private final String jsonTable;
+      private final String jsonSchema;
+      private final WriteDisposition writeDisposition;
+      private final CreateDisposition createDisposition;
+      private final Coder<T> coder;
+
+      private BigQueryWriteOperation(
+          BigQuerySink<T> sink,
+          String jobIdToken,
+          String jsonTable,
+          String jsonSchema,
+          WriteDisposition writeDisposition,
+          CreateDisposition createDisposition,
+          Coder<T> coder) {
+        super(sink);
+        this.jobIdToken = jobIdToken;
+        this.jsonTable = jsonTable;
+        this.jsonSchema = jsonSchema;
+        this.writeDisposition = writeDisposition;
+        this.createDisposition = createDisposition;
+        this.coder = coder;
+      }
+
+      @Override
+      public FileBasedWriter<T> createWriter(PipelineOptions options) throws Exception {
+        return new TextWriter<>(this, coder);
+      }
+
+      @Override
+      public void finalize(Iterable<FileResult> writerResults, PipelineOptions options)
+          throws Exception {
+        BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+        Bigquery client = Transport.newBigQueryClient(bqOptions).build();
+        BigQueryTableInserter inserter = new BigQueryTableInserter(client);
+        List<String> tempFiles = Lists.newArrayList();
+        for (FileResult result : writerResults) {
+          tempFiles.add(result.getFilename());
+        }
+        if (!tempFiles.isEmpty()) {
+          inserter.load(
+              jobIdToken,
+              JSON_FACTORY.fromString(jsonTable, TableReference.class),
+              tempFiles,
+              JSON_FACTORY.fromString(jsonSchema, TableSchema.class),
+              writeDisposition,
+              createDisposition);
+        }
+      }
+    }
+
+    private static class TextWriter<T> extends FileBasedWriter<T> {
+      private static final byte[] NEWLINE = "\n".getBytes(StandardCharsets.UTF_8);
+      private final Coder<T> coder;
+      private OutputStream out;
+
+      public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder) {
+        super(writeOperation);
+        this.mimeType = MimeTypes.TEXT;
+        this.coder = coder;
+      }
+
+      @Override
+      protected void prepareWrite(WritableByteChannel channel) throws Exception {
+        out = Channels.newOutputStream(channel);
+      }
+
+      @Override
+      public void write(T value) throws Exception {
+        coder.encode(value, out, Context.OUTER);
```

Comment:
It uses newline as the delimeter.
Otherwise, the outside system won't understand the delimeter of Context.NESTED.